File: Algorithms/Backtracking/0098_validate_BST.c
// Practice for: 0098_validate_BST.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>
#include <limits.h> 

struct TreeNode {
    int val; 
    struct TreeNode *left;
    struct TreeNode *right;
}; 

bool validate(struct TreeNode* node, long long min_val, long long max_val) {
    if (node == NULL){
        return true;
    }

    //fail condition
    if (node->val <= min_val || node->val >= max_val){
        return false;
    }

    return validate(node->left, min_val, node->val) && validate(node->right, node->val, max_val);

}

bool isValidBST(struct TreeNode *root){
    return validate(root, LLONG_MIN, LLONG_MAX);
}

File: Algorithms/Backtracking/0098_validate_BST.py
#backtracking

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: bool
        """
        def valid(node, minimum, maximum):
            if not node: 
                return True
            
            if not (node.val > minimum and node.val < maximum):
                return False
            #left must be smaller, right must be bigger 
            return valid(node.left, minimum, node.val) and valid(node.right, node.val, maximum)
            
        return valid (root, float("-inf"), float("inf"))
        